### PR TITLE
Big refactor to clean up the WatchSlots logic

### PR DESCRIPTION
**Pull request by johnstonematt**
## Main Refactor
I replaced the monolithic `WatchSlots()` with the `SolanaCollector.WatchSlots()`, which I personally believe is much better for reasons such as:
* Clear indication of state and tracking parameters
* More, smaller functions

In general, the logic in the previous one was just insanely hard to follow, and I am 99% certain there were double counting errors on epoch boundaries. 

One of the main things I focused on in `SolanaCollector.WatchSlots()` was to clearly mark the logic that is executed on epoch boundary, to make it easy to implement things like inflationary and Jito reward tracking, which happens at epoch bounds.

## Other changes:
* I refactored the `rpc.Client` to make the functions more like a direct overlay of the [RPC API](https://solana.com/docs/rpc/http), without any additional logic.
* Updated the `Commitment` levels, as the ones used have been deprecated for a while.

## Still to come:
With the logic split up more granularly now, more tests can be done.

### CODE REVIEW

File: cmd/solana_exporter/slots.go
Comment (johnstonematt) [inline cmd/solana_exporter/slots.go:1]: I would advise just reading this file plainly as is, as the whole thing was rewritten and so the diff is not very helpful IMO

File: cmd/solana_exporter/slots.go [hunk 1]
```diff
@@ -68,182 +87,161 @@ func init() {
...
+// fetchAndEmitBlockProduction fetches block production up to the provided endSlot, emits the prometheus metrics,
+// and updates the SlotWatcher.slotWatermark accordingly
+func (c *SlotWatcher) fetchAndEmitBlockProduction(ctx context.Context, endSlot int64) {
+	// add 1 because GetBlockProduction's range is inclusive, and the watermark is already tracked
+	startSlot := c.slotWatermark + 1
+	klog.Infof("Fetching block production in [%v -> %v]", startSlot, endSlot)
+
+	// make sure the bounds are contained within the epoch we are currently watching:
+	if err := c.checkValidSlotRange(startSlot, endSlot); err != nil {
+		panic(err)
```
Comment (SEJeff) [inline cmd/solana_exporter/slots.go:211]: Is a panic really necessary in this case? Can we not just gracefully skip this or reset things some how? Also, perhaps instead of a panic (which dumps all goroutines + stack traces to stdout, we just do something more like [klog.Fatal(err.String())](https://pkg.go.dev/k8s.io/klog/v2#Fatal) or maybe [klog.Fatalf("slot range is out of bounds of the current epoch: %v", err)](https://pkg.go.dev/k8s.io/klog/v2#Fatalf)

File: cmd/solana_exporter/utils.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+		panic(fmt.Errorf(format, args...))
+	}
+}
+
+// prettyPrint is just a useful debugging function
+func prettyPrint(obj any) {
+	// For pretty-printed JSON, use json.MarshalIndent
+	prettyJSON, err := json.MarshalIndent(obj, "", "  ")
+	if err != nil {
+		fmt.Println("Error marshalling to pretty JSON:", err, ". obj: ", obj)
```
Comment (SEJeff) [inline cmd/solana_exporter/utils.go:19]: Should this be using something like this? [klog.Info("error marshalling to pretty JSON: %s. obj: %v", err, obj)](https://pkg.go.dev/k8s.io/klog/v2#Infof)

Bare fmt.Print.. in code that is long running is generally a bug.

File: pkg/rpc/client.go [hunk 3]
```diff
@@ -70,23 +72,22 @@ func (c Commitment) MarshalJSON() ([]byte, error) {
...
-	CommitmentRoot Commitment = "root"
-	// CommitmentSingleGossip represents the most recent block that has been voted on
-	//by the cluster super-majority (optimistic confirmation).
-	CommitmentSingleGossip Commitment = "singleGossip"
-	// CommitmentRecent represents the nodes most recent block
-	CommitmentRecent Commitment = "recent"
+	// CommitmentFinalized level offers the highest level of certainty for a transaction on the Solana blockchain.
+	// A transaction is considered “Finalized” when it is included in a block that has been confirmed by a
+	// supermajority of the stake, and at least 31 additional confirmed blocks have been built on top of it.
+	CommitmentFinalized Commitment = "finalized"
```
Comment (SEJeff) [inline pkg/rpc/client.go:78]: For the usage of the exporter, do we actually need to use finalized blocks? This is ~15 seconds behind the latest roughly. For monitoring, max is probably good enough.

File: pkg/rpc/client.go [hunk 4]
```diff
@@ -149,47 +158,61 @@ func (c *Client) GetVersion(ctx context.Context) (string, error) {
...
+	if identity != nil {
+		config["identity"] = *identity
+	}
 	if firstSlot != nil {
-		params[0] = map[string]interface{}{"range": blockProductionRange{FirstSlot: *firstSlot, LastSlot: lastSlot}}
+		blockRange := map[string]int64{"firstSlot": *firstSlot}
+		if lastSlot != nil {
+			// make sure first and last slot are in order:
+			if *firstSlot > *lastSlot {
+				panic(fmt.Errorf("last slot %v is greater than first slot %v", *lastSlot, *firstSlot))
```
Comment (SEJeff) [inline pkg/rpc/client.go:193]: Can we just return this as an error instead of panicking? Less panicking is good.

**Review by SEJeff — APPROVED** [review]

File: cmd/solana_exporter/slots.go (hunk 1 above)
Comment (johnstonematt) [reply cmd/solana_exporter/slots.go:211]: The reason I panic here is because if it happens then it's a logic issue in the code, not just some random event.
